tests/trx_generator/trx_provider.hpp

```diff
@@ -23,40 +23,82 @@ namespace eosio::testing {
    };
 
    struct provider_base_config {
-      std::string _peer_endpoint = "127.0.0.1";
-      unsigned short _port = 9876;
+      std::string    _peer_endpoint_type = "p2p";
+      std::string    _peer_endpoint      = "127.0.0.1";
+      unsigned short _port               = 9876;
 
       std::string to_string() const {
          std::ostringstream ss;
-         ss << "peer_endpoint: " << _peer_endpoint << " port: " << _port;
+         ss << "endpoint type: " << _peer_endpoint_type << "peer_endpoint: " << _peer_endpoint << " port: " << _port;
          return std::move(ss).str();
       }
    };
 
-   struct p2p_connection {
-      const provider_base_config& _config;
-      boost::asio::io_service _p2p_service;
-      boost::asio::ip::tcp::socket _p2p_socket;
+   struct provider_connection {
+      const provider_base_config&                                 _config;
+      eosio::chain::named_thread_pool<struct provider_connection> _connection_thread_pool;
+
+      provider_connection(const provider_base_config& provider_config)
+          : _config(provider_config) {}
+
+      void init_and_connect() {
+         _connection_thread_pool.start(
+             1, [](const fc::exception& e) { elog("provider_connection exception ${e}", ("e", e)); });
+         connect();
+      };
+
+      void cleanup_and_disconnect() {
+         disconnect();
+         _connection_thread_pool.stop();
+      };
+
+      virtual void send_transaction(const chain::packed_transaction& trx) = 0;
+
+    private:
+      virtual void connect()    = 0;
+      virtual void disconnect() = 0;
+   };
+
+   struct http_connection : public provider_connection {
+      uint64_t _acknowledged = 0;
+      uint64_t _sent         = 0;
```

Comment:
These should be `std::atomic<uint64_t>` since accessed/modified from different threads.

Reply:
Good call. Being currently single threaded I hadn't attacked that.  Addressed: [f342f5e](https://github.com/AntelopeIO/leap/pull/1138/commits/f342f5e3cace6bd7db29e548640fc9a23fe53d43)